lib/puppet/functions/resources_deep_merge.rb

```diff
@@ -0,0 +1,87 @@
+# @summary Deeply merge a "defaults" hash into a "resources" hash like the ones expected by `create_resources()`.
+#
+# Deeply merge a "defaults" hash into a "resources" hash like the ones expected by `create_resources()`.
+#
+# Internally calls the puppetlabs-stdlib function `deep_merge()`. In case of
+# duplicate keys the `resources` hash keys win over the `defaults` hash keys.
+#
+# Example
+# ```puppet
+# $defaults_hash = {
+#   'one'   => '1',
+#   'two'   => '2',
+#   'three' => '3',
+#   'four'  => {
+#     'five'  => '5',
+#     'six'   => '6',
+#     'seven' => '7',
+#   }
+# }
+#
+# $numbers_hash = {
+#   'german' => {
+#     'one'   => 'eins',
+#     'three' => 'drei',
+#     'four'  => {
+#       'six' => 'sechs',
+#     },
+#   },
+#   'french' => {
+#     'one' => 'un',
+#     'two' => 'deux',
+#     'four' => {
+#       'five'  => 'cinq',
+#       'seven' => 'sept',
+#     },
+#   }
+# }
+#
+# $result_hash = resources_deep_merge($numbers_hash, $defaults_hash)
+# ```
+#
+# The $result_hash then looks like this:
+#
+# ```puppet
+# $result_hash = {
+#   'german' => {
+#     'one'   => 'eins',
+#     'two'   => '2',
+#     'three' => 'drei',
+#     'four'  => {
+#       'five'  => '5',
+#       'six'   => 'sechs',
+#       'seven' => '7',
+#     }
+#   },
+#   'french' => {
+#     'one'   => 'un',
+#     'two'   => 'deux',
+#     'three' => '3',
+#     'four'  => {
+#       'five'  => 'cinq',
+#       'six'   => '6',
+#       'seven' => 'sept',
+#     }
+#   }
+# }
+# ```
+Puppet::Functions.create_function(:resources_deep_merge) do
+  # Deep-merges defaults into a resources hash
+  # @param resources The hash of resources.
+  # @param defaults The hash of defaults to merge.
+  # @return [Hash] Returns the merged hash.
```

Comment:
The `[Hash]` part is also added by puppet-strings for functions which have a return type.

Reply:
Right, sorry.  You only need to specify the type here if there are multiple possible return types you're documenting (and that 'feature' doesn't work yet anyway!).

Reply:
updated